File: src/SubscriptionManagement/CreateSubscription.cs
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;
using Amazon.Lambda.Core;
using Amazon.Lambda.APIGatewayEvents;
using System;
using Newtonsoft.Json;
using Microsoft.Extensions.DependencyInjection;
using Amazon.DynamoDBv2;
using Amazon.DynamoDBv2.Model;

namespace CatchTheRun.SubscriptionManagement
{
    public class CreateSubscription
    {
        private readonly IServiceProvider serviceProvider;
        public CreateSubscription() : this(SubscriptionManagementConfiguration.BuildContainer().BuildServiceProvider())
        {
        }

        public CreateSubscription(IServiceProvider serviceProvider)
        {
            this.serviceProvider = serviceProvider;
        }

        [LambdaSerializer(typeof(Amazon.Lambda.Serialization.Json.JsonSerializer))]
        public async Task<APIGatewayProxyResponse> Execute(APIGatewayProxyRequest req, ILambdaContext ctx)
        {
            ctx.Logger.LogLine($"Lambda Context: {JsonConvert.SerializeObject(ctx)}");
            ctx.Logger.LogLine($"Create Subscription Request: {JsonConvert.SerializeObject(req)}");

            var reqBody = JsonConvert.DeserializeObject<CreateSubscriptionRequestBody>(req.Body);

            using (var DynamoClient = serviceProvider.GetService<IAmazonDynamoDB>())
            {
                var key = new Dictionary<string, AttributeValue>()
                {
                    { "PRT", new AttributeValue() { S = reqBody.ProviderName } },
                    { "SRT", new AttributeValue() { S = $"F|SUB|{reqBody.ConsumerName}"} }
                };
                var getSubResponse = await DynamoClient.GetItemAsync("Main", key);
                ctx.Logger.LogLine($"Get Sub Response: {JsonConvert.SerializeObject(getSubResponse)}");

                if (!getSubResponse.IsItemSet)
                {
                    var attributes = new Dictionary<string, AttributeValue>();
                    attributes["PRT"] = new AttributeValue() { S = reqBody.ProviderName };
                    attributes["SRT"] = new AttributeValue() { S = $"F|SUB|{reqBody.ConsumerName}" };
                    attributes["G1S"] = new AttributeValue() { S = reqBody.ProviderName };
                    attributes[reqBody.Protocol] = new AttributeValue()
                    {
                        M = new Dictionary<string, AttributeValue>()
                        {
                            { "Endpoint", new AttributeValue() { S = reqBody.Endpoint } },
                            { "DiscordServerId", new AttributeValue() { S = reqBody.DiscordServerId } },
                            { "FilterPolicy", new AttributeValue()
                                {
                                    M = new Dictionary<string, AttributeValue>()
                                    {
                                        { "PolicyType", new AttributeValue() { S = reqBody.FilterPolicy.Type } },
                                        { "Games", new AttributeValue() { SS = reqBody.FilterPolicy.Games } },
                                        { "Categories", new AttributeValue() { SS = reqBody.FilterPolicy.Categories } }
                                    }
                                }
                            }
                        }
                    };
                    var createSubRequest = new PutItemRequest() { TableName = "Main", Item = attributes };
                    var createSubResponse = await DynamoClient.PutItemAsync(createSubRequest);
                    return new APIGatewayProxyResponse() { StatusCode = 200, Headers = new Dictionary<string, string> { { "Access-Control-Allow-Origin", "*" } } };
                }
                else
                {
                    var sub = getSubResponse.Item;
                    sub[reqBody.Protocol] = new AttributeValue()
                    {
                        M = new Dictionary<string, AttributeValue>()
                        {
                            { "Endpoint", new AttributeValue() { S = reqBody.Endpoint } },
                            { "DiscordServerId", new AttributeValue() { S = reqBody.DiscordServerId } },
                            { "FilterPolicy", new AttributeValue()
                                {
                                    M = new Dictionary<string, AttributeValue>()
                                    {
                                        { "PolicyType", new AttributeValue() { S = reqBody.FilterPolicy.Type } },
                                        { "Games", new AttributeValue() { SS = reqBody.FilterPolicy.Games } },
                                        { "Categories", new AttributeValue() { SS = reqBody.FilterPolicy.Categories } }
                                    }
                                }
                            }
                        }
                    };
                    var addSubInstanceRequest = new PutItemRequest() { TableName = "Main", Item = sub };
                    var addSubInstanceResponse = await DynamoClient.PutItemAsync(addSubInstanceRequest);
                    return new APIGatewayProxyResponse() { StatusCode = 200, Headers = new Dictionary<string, string> { { "Access-Control-Allow-Origin", "*" } } };
                }
            }
        }
    }
}

File: subscriptionManagement/createSubscription.js
"use strict";

const dynamoDB = require("aws-sdk/clients/dynamodb");
const dynamoClient = new dynamoDB.DocumentClient();

module.exports.createSubscription = async (event, context) => {
  console.log("Lambda Context:", context);
  console.log("Create Subscription Request:", event);

  const newSubInstance = {
    Endpoint: event.Endpoint,
    DiscordServerId: event.DiscordServerId,
    FilterPolicy: event.FilterPolicy
  };

  const params = {
    TableName: "Main",
    Key: {
      PRT: event.ProviderName,
      SRT: `F|SUB|${event.ConsumerName}`
    }
  };

  let getSubResponse;

  try {
    getSubResponse = await dynamoClient.get(params).promise();
  } catch (e) {
    console.log("error:", e);
  }

  if (getSubResponse.Item) {
    const existingSub = getSubResponse.Item;
    existingSub[event.Protocol] = newSubInstance;

    try {
      const addSubInstanceResponse = await dynamoClient
        .put({ TableName: "Main", Item: existingSub })
        .promise();
      return {
        statusCode: 200,
        headers: {
          "Access-Control-Allow-Origin": "*"
        }
      };
    } catch (e) {
      console.log("error:", e);
    }
  } else {
    const newSub = {
      PRT: event.ProviderName,
      SRT: `F|SUB|${event.ConsumerName}`,
      G1S: event.ProviderName
    };
    newSub[event.Protocol] = newSubInstance;

    try {
      const addNewSubResponse = await dynamoClient
        .put({ TableName: "Main", Item: newSub })
        .promise();
      return {
        statusCode: 200,
        headers: {
          "Access-Control-Allow-Origin": "*"
        }
      };
    } catch (e) {
      console.log("error:", e);
    }
  }
};
